File: Sources/TimeoutEngine/ConfigStore.swift
import Foundation

/// 配置与引擎状态的 JSON 持久化（纯 Foundation，可单测）。
/// 路径：~/Library/Application Support/<bundleId>/{config.json, engine-state.json}
/// 原子写 + schema 版本迁移；读取失败回退默认/nil，绝不抛出到调用方。
public final class ConfigStore {
    private let directory: URL
    private let fileManager = FileManager.default

    public init(directory: URL) throws {
        self.directory = directory
        try fileManager.createDirectory(at: directory, withIntermediateDirectories: true)
    }

    /// 默认目录：~/Library/Application Support/<bundleId>。
    public static func defaultDirectory(bundleId: String) -> URL {
        let support = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? URL(fileURLWithPath: NSTemporaryDirectory())
        return support.appendingPathComponent(bundleId, isDirectory: true)
    }

    private var configURL: URL { directory.appendingPathComponent("config.json") }
    private var stateURL: URL { directory.appendingPathComponent("engine-state.json") }

    // MARK: - Config

    public func loadConfig() -> DayPlanConfig {
        guard let data = try? Data(contentsOf: configURL) else { return .defaultConfig }
        do {
            let decoded = try JSONDecoder().decode(DayPlanConfig.self, from: data)
            return migrate(decoded)
        } catch {
            NSLog("[Timeout] config 解码失败，回退默认：\(error)")
            return .defaultConfig
        }
    }

    public func saveConfig(_ config: DayPlanConfig) throws {
        try write(config, to: configURL)
    }

    // MARK: - Engine State（崩溃恢复用）

    public func loadState() -> EngineState? {
        guard let data = try? Data(contentsOf: stateURL) else { return nil }
        return try? JSONDecoder().decode(EngineState.self, from: data)
    }

    public func saveState(_ state: EngineState) {
        do {
            try write(state, to: stateURL)
        } catch {
            NSLog("[Timeout] 引擎状态持久化失败：\(error)")
        }
    }

    // MARK: - Private

    private func write<T: Encodable>(_ value: T, to url: URL) throws {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        let data = try encoder.encode(value)
        try data.write(to: url, options: .atomic)  // 原子写，防半写损坏
    }

    /// schema 迁移：未来版本号升级时在此逐级迁移；高于本版本则回退默认。
    private func migrate(_ config: DayPlanConfig) -> DayPlanConfig {
        guard config.schemaVersion <= DayPlanConfig.currentSchemaVersion else {
            NSLog("[Timeout] config schemaVersion(\(config.schemaVersion)) 高于本版本(\(DayPlanConfig.currentSchemaVersion))，回退默认")
            return .defaultConfig
        }
        return config
    }
}

File: windows/GiveMeABreakEngine/ConfigStore.cs
using System.Text.Json;

namespace GiveMeABreakEngine;

// MARK: - 配置与引擎状态 JSON 持久化
// 镜像 Sources/GiveMeABreakEngine/ConfigStore.swift。原子写 + schema 迁移；读取失败回退默认/nil，绝不抛出。

public sealed class ConfigStore
{
    private readonly string _directory;

    public ConfigStore(string directory)
    {
        _directory = directory;
        Directory.CreateDirectory(_directory);
    }

    /// <summary>默认目录：%APPDATA%\&lt;bundleId&gt;（Windows）。Phase 1+ 集成层使用。</summary>
    public static string DefaultDirectory(string bundleId)
        => Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), bundleId);

    private string ConfigPath => Path.Combine(_directory, "config.json");
    private string StatePath => Path.Combine(_directory, "engine-state.json");

    public DayPlanConfig LoadConfig()
    {
        if (!File.Exists(ConfigPath)) return DayPlanConfig.Default;
        try
        {
            var json = File.ReadAllText(ConfigPath);
            return Migrate(DayPlanConfig.FromJson(json, GiveMeABreakJsonOptions.Default));
        }
        catch
        {
            return DayPlanConfig.Default;
        }
    }

    public void SaveConfig(DayPlanConfig config)
        => WriteAtomic(ConfigPath, JsonSerializer.Serialize(config, GiveMeABreakJsonOptions.Default));

    public EngineState? LoadState()
    {
        if (!File.Exists(StatePath)) return null;
        try
        {
            return JsonSerializer.Deserialize<EngineState>(File.ReadAllText(StatePath), GiveMeABreakJsonOptions.Default);
        }
        catch
        {
            return null;
        }
    }

    public void SaveState(EngineState state)
    {
        try { WriteAtomic(StatePath, JsonSerializer.Serialize(state, GiveMeABreakJsonOptions.Default)); }
        catch { /* 静默：镜像 Swift NSLog 但不抛出到调用方 */ }
    }

    private static void WriteAtomic(string path, string content)
    {
        // 原子写：先写临时文件再覆盖，防半写损坏（对齐 Swift data.write(options: .atomic)）。
        var tmp = path + ".tmp";
        File.WriteAllText(tmp, content);
        File.Move(tmp, path, overwrite: true);
    }

    /// <summary>schema 迁移：高于本版本则回退默认；否则规范化 schemaVersion 为当前版本
    /// （FromJson 已对缺失字段容错补默认，故此处无需逐字段迁移分支）。</summary>
    private static DayPlanConfig Migrate(DayPlanConfig config)
    {
        if (config.SchemaVersion > DayPlanConfig.CurrentSchemaVersion) return DayPlanConfig.Default;
        var c = config;
        c.SchemaVersion = DayPlanConfig.CurrentSchemaVersion;  // 规范化，下次落盘即升级
        return c;
    }
}

/// <summary>共享 JSON 选项：camelCase（对齐 Swift Codable 默认形态）+ 缩进。</summary>
public static class GiveMeABreakJsonOptions
{
    public static readonly JsonSerializerOptions Default = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        WriteIndented = true,
    };
}
